### ISSUE TITLE
Running into problems with mmcv==1.3.18

### ISSUE THREAD

**Issue by jfitzg7** [YangHai-1218/RADet#6]
Hello,

I'm trying to run the training script in tools/train.py and I ran into a few errors with mmcv (using version 1.3.18).

The first problem I ran into complained about `mmcv._ext`:

```
$ python tools/train.py --config configs/bop/r50_ycbv_pbr_testing123.py           
Traceback (most recent call last):                                                                                                                            
  File "/data2/6d-pose-estimation/radet-attempt-2/RADet/tools/train.py", line 15, in <module>                                                                 
    from radet.apis import set_random_seed, train_detector                                                                                                    
  File "/data2/6d-pose-estimation/radet-attempt-2/RADet/radet/apis/__init__.py", line 1, in <module>                                                          
    from .inference import (async_inference_detector, inference_detector,                                                                                     
  File "/data2/6d-pose-estimation/radet-attempt-2/RADet/radet/apis/inference.py", line 6, in <module>                                                         
    from mmcv.ops import RoIPool                                                                                                                              
  File "/home/exx/anaconda3/envs/radet-attempt-2-0/lib/python3.9/site-packages/mmcv/ops/__init__.py", line 2, in <module>                                     
    from .assign_score_withk import assign_score_withk                                                                                                        
  File "/home/exx/anaconda3/envs/radet-attempt-2-0/lib/python3.9/site-packages/mmcv/ops/assign_score_withk.py", line 5, in <module>                           
    ext_module = ext_loader.load_ext(                                                                                                                         
  File "/home/exx/anaconda3/envs/radet-attempt-2-0/lib/python3.9/site-packages/mmcv/utils/ext_loader.py", line 13, in load_ext                                
    ext = importlib.import_module('mmcv.' + name)                                                                                                             
  File "/home/exx/anaconda3/envs/radet-attempt-2-0/lib/python3.9/importlib/__init__.py", line 127, in import_module                                           
    return _bootstrap._gcd_import(name[level:], package, level)                                                                                               
ModuleNotFoundError: No module named 'mmcv._ext' 
```

But I was able to find this issue on GitHub (https://github.com/open-mmlab/mmcv/issues/204) that provided a solution for it which basically boiled down to running `MMCV_WITH_OPS=1 pip install -e .` in the root directory of the source code for mmcv==1.3.18. But then I ran into another issue:

```
$ python tools/train.py --config configs/bop/r50_ycbv_pbr_testing123.py 
Traceback (most recent call last):
  File "/data2/6d-pose-estimation/radet-attempt-2/RADet/tools/train.py", line 15, in <module>
    from radet.apis import set_random_seed, train_detector
  File "/data2/6d-pose-estimation/radet-attempt-2/RADet/radet/apis/__init__.py", line 1, in <module>
    from .inference import (async_inference_detector, inference_detector,
  File "/data2/6d-pose-estimation/radet-attempt-2/RADet/radet/apis/inference.py", line 6, in <module>
    from mmcv.ops import RoIPool
  File "/home/exx/anaconda3/envs/radet-attempt-2-0/lib/python3.9/site-packages/mmcv-1.3.18/mmcv/ops/__init__.py", line 2, in <module>
    from .assign_score_withk import assign_score_withk
  File "/home/exx/anaconda3/envs/radet-attempt-2-0/lib/python3.9/site-packages/mmcv-1.3.18/mmcv/ops/assign_score_withk.py", line 5, in <module>
    ext_module = ext_loader.load_ext(
  File "/home/exx/anaconda3/envs/radet-attempt-2-0/lib/python3.9/site-packages/mmcv-1.3.18/mmcv/utils/ext_loader.py", line 13, in load_ext
    ext = importlib.import_module('mmcv.' + name)
  File "/home/exx/anaconda3/envs/radet-attempt-2-0/lib/python3.9/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
ImportError: /home/exx/anaconda3/envs/radet-attempt-2-0/lib/python3.9/site-packages/mmcv-1.3.18/mmcv/_ext.cpython-39-x86_64-linux-gnu.so: undefined symbol: _Z27points_in_boxes_cpu_forwardN2at6TensorES0_S0_
```

I found this GitHub issue (https://github.com/open-mmlab/mmcv/issues/1556) from the mmcv repository that claims this problem was fixed in mmcv==1.4.0

My question is, are you guys sure that mmcv==1.3.18 is the right version to be using? Am I not installing it correctly? Would it be preferable to use a version lower than 1.3.18, such as 1.3.17, or would it be better to use 1.4.0?

**Comment by YangHai-1218** [comment]
Yes, i can confirm we are using mmcv==1.3.18. How do you install mmcv? Using the pre-built mmcv-full package is  okay.

**Comment by jfitzg7** [comment]
Hi @YangHai-1218, thank you for your response!

Essentially the steps for what I'm doing to install is: 

1. downloading mmcv from PyPI (https://pypi.org/project/mmcv/1.3.18/#files) via:
```
(radet-attempt-2-0) ~/anaconda3/envs/radet-attempt-2-0/lib/python3.9/site-packages$ wget https://files.pythonhosted.org/packages/c4/b5/7b1e9cb7a80c6d64843d39627512ecd19fa3d75ffa4ab3e5bf1570b13d64/mmcv-1.3.18.tar.gz
```
2. Extract the source code via:
```
(radet-attempt-2-0) ~/anaconda3/envs/radet-attempt-2-0/lib/python3.9/site-packages$ tar -xvzf mmcv-1.3.18.tar.gz
```
3. move into the directory via:
```
(radet-attempt-2-0) ~/anaconda3/envs/radet-attempt-2-0/lib/python3.9/site-packages$ cd mmcv-1.3.18
```
4. Then install the package with pip via:
```
(radet-attempt-2-0) ~/anaconda3/envs/radet-attempt-2-0/lib/python3.9/site-packages/mmcv-1.3.18$ MMCV_WITH_OPS=1 pip install -e .
```

Some important things to note:
* I am on `ubuntu 20.04`
* I am using Anaconda to manage the virtual environment `conda --version` outputs => `conda 4.10.3`
* I am using python 3.9, and I started with a bare conda environment by running `conda create --name radet-attempt-2-0 python=3.9`
* My $CUDA_HOME environment variable contains `/usr/local/cuda-11.1` (which I believe is used to build mmcv)
* My `/usr/local/cuda` symlink points to => `/usr/local/cuda-11.3`
* I am using torch==1.10.1+cu111 and torchvision=0.11.2+cu111 which I installed via the command:
```
(radet-attempt-2-0) ~$ pip install torch==1.10.1+cu111 torchvision==0.11.2+cu111 -f https://download.pytorch.org/whl/cu111/torch_stable.html
```

I think the torch/torchvision versions and the cuda version (specified by the $CUDA_HOME variable which contains `/usr/local/cuda-11.1`) could be a problem since they don't line up with the requirements.txt and it's using `cuda-11.1` and not `cuda-11.3`. However, I ran into problems with trying to install torch==1.10.0 and torchvision==0.11.0 by getting the following error:
```
(radet-attempt-2-0) ~$ pip install torch==1.10.0 torchvision==0.11.0
Collecting torch==1.10.0
  Using cached torch-1.10.0-cp39-cp39-manylinux1_x86_64.whl (881.9 MB)
Collecting torchvision==0.11.0
  Using cached torchvision-0.11.0-cp39-cp39-manylinux1_x86_64.whl (23.2 MB)
Requirement already satisfied: typing-extensions in /home/exx/anaconda3/envs/radet-attempt-2-0/lib/python3.9/site-packages (from torch==1.10.0) (4.7.1)
Requirement already satisfied: numpy in /home/exx/anaconda3/envs/radet-attempt-2-0/lib/python3.9/site-packages (from torchvision==0.11.0) (1.25.2)
INFO: pip is looking at multiple versions of torchvision to determine which version is compatible with other requirements. This could take a while.
ERROR: Cannot install torch==1.10.0 and torchvision==0.11.0 because these package versions have conflicting dependencies.

The conflict is caused by:
    The user requested torch==1.10.0
    torchvision 0.11.0 depends on torch==1.10.0+cu102

To fix this you could try to:
1. loosen the range of package versions you've specified
2. remove package versions to allow pip attempt to solve the dependency conflict

ERROR: ResolutionImpossible: for help visit https://pip.pypa.io/en/latest/topics/dependency-resolution/#dealing-with-dependency-conflicts

```

Is it possible that I need to use cuda-10.2 for the toolkit version? Could `ubuntu 20.04` be the problem?

Also I tested it with mmcv==1.3.17 and I did not run into that second error that I original shared, but I then ran into an issue with mmdet:
```
(radet-attempt-2-0) /data2/6d-pose-estimation/radet-attempt-2/RADet$ python tools/train.py --config configs/bop/r50_ycbv_pbr_testing123.py           
Traceback (most recent call last):                                                                                                                            
  File "/data2/6d-pose-estimation/radet-attempt-2/RADet/tools/train.py", line 15, in <module>                                                                 
    from radet.apis import set_random_seed, train_detector                                                                                                    
  File "/data2/6d-pose-estimation/radet-attempt-2/RADet/radet/apis/__init__.py", line 1, in <module>                                                          
    from .inference import (async_inference_detector, inference_detector,                                                                                     
  File "/data2/6d-pose-estimation/radet-attempt-2/RADet/radet/apis/inference.py", line 10, in <module>                                                        
    from radet.core import get_classes                                                                                                                        
  File "/data2/6d-pose-estimation/radet-attempt-2/RADet/radet/core/__init__.py", line 2, in <module>                                                          
    from .bbox import *  # noqa: F401, F403                                                                                                                   
  File "/data2/6d-pose-estimation/radet-attempt-2/RADet/radet/core/bbox/__init__.py", line 7, in <module>                                                     
    from .samplers import (BaseSampler, CombinedSampler,                                                                                                      
  File "/data2/6d-pose-estimation/radet-attempt-2/RADet/radet/core/bbox/samplers/__init__.py", line 1, in <module>                                            
    from .base_sampler import BaseSampler                                                                                                                     
  File "/data2/6d-pose-estimation/radet-attempt-2/RADet/radet/core/bbox/samplers/base_sampler.py", line 5, in <module>                                        
    from .sampling_result import SamplingResult                                                                                                               
  File "/data2/6d-pose-estimation/radet-attempt-2/RADet/radet/core/bbox/samplers/sampling_result.py", line 3, in <module>                                     
    from mmdet.utils import util_mixins                                                                                                                       
ModuleNotFoundError: No module named 'mmdet'
```

Is mmdet supposed to be a dependency? If so, do you guys know which version we should be using? I do not currently have mmdet installed in my radet-attempt-2-0 virtual environment, and I also do not see an entry for it in the requirements.txt.

**Comment by YangHai-1218** [comment]
First, you should check out the mmcv official instructions for installing mmcv, and directly install the pre-built packages provided by mmcv.
Second, the error about dependency for mmdet has been fixed. Please pull the latest code and try again.

**Comment by jfitzg7** [comment]
Thanks @YangHai-1218! I think installing the pre-built version did the trick! I removed the old mmcv and mmcv-full packages (I had to run `rm -rf` on the mmcv package folder in the site-packages to actually remove it). and then I used the following commands:
```
pip install -U openmim
mim install mmcv-full==1.3.18
```
And it did seem to actually download the pre-built version directly, as indicated in their subsection "Installation log using pre-built packages". It did not download the .tar.gz and build from source, which I'm guessing is the critical part. But to be fair I haven't tested to see if it would work if it took the "build from source" route instead (although maybe that was what I was technically doing before?)

I did run into a few more errors with the training, however I believe they are unrelated to mmcv. I think you have fixed all the problems that I was having in this issue! Thank you! I'm going to open a new issue related to the problems I'm seeing now If I can't figure this out. (I think I'm not creating the annotations correctly)
